src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -72,4 +92,203 @@ object RuleUtils {
+   *   - if Project filters indexedColumns, then Shuffle should be located after the node
+   *     - Project => Shuffle => Filter => Relation
+   * 3) Bucket-aware union both indexPlan and complementPlan to avoid repartitioning index data
+   *
+   * NOTE: This method currently only supports replacement of Scan Nodes i.e. Logical relations
+   *
+   * @param spark Spark session
+   * @param index index used in replacement plan
+   * @param originalPlan original plan
+   * @param indexPlan replaced plan with index
+   * @return complementIndexPlan integrated plan of indexPlan and complementPlan
+   */
+  private def getComplementIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      originalPlan: LogicalPlan,
+      indexPlan: LogicalPlan): LogicalPlan = {
+    // 1) Replace the location of LogicalRelation with appended files
+    val complementIndexPlan = originalPlan transformUp {
+      case baseRelation @ LogicalRelation(
+            fsRelation @ HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
+            baseOutput,
+            _,
+            _) =>
+        // Set the same output schema with the index plan to merge them using BucketUnion
+        val updatedOutput =
+          baseOutput.filter(attr => index.schema.fieldNames.contains(attr.name))
+
+        val filesNotCovered = (location.allFiles
+          .map(f => FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
+          .toSet -- index.allSourceFiles).toSeq.map(f => new Path(f.name))
+
+        if (filesNotCovered.nonEmpty) {
+          val newLocation =
+            new InMemoryFileIndex(spark, filesNotCovered, Map(), None)
+          val newRelation =
+            fsRelation.copy(
+              location = newLocation,
+              dataSchema = StructType(index.schema.filter(baseRelation.schema.contains(_))))(
+              spark)
+          baseRelation.copy(relation = newRelation, output = updatedOutput)
+        } else {
+          baseRelation
+        }
+    }
+
+    if (!originalPlan.equals(complementIndexPlan)) {
+      // Remove sort order because we cannot guarantee the ordering of source files
+      val bucketSpec = index.bucketSpec.copy(sortColumnNames = Seq())
+      val attrs = complementIndexPlan.output.attrs.filter { attr =>
+        index.indexedColumns contains attr.name
+      }
+
+      // 2) Perform on-the-fly shuffle with the same partition structure of index
+      // so that we could avoid incurring shuffle whole index data at merge stage.
+      val shuffled = if (attrs.size != index.indexedColumns.size) {
+        // There should be a project node which excludes one or more indexedColumns.
+        // In this case, shuffle will be replaced to RoundRobinPartitioning which can cause
+        // wrong result issue. So shuffle should be located before the project node.
+        var projectSeen = false
+        val transformed = complementIndexPlan transformUp {
```

Comment:
This will add RepartitionByExpression after every Project node in this subplan. That's ok right? I am not sure if this can cause issues for e.g. if the plan is 
`Project -> SomeNode -> Project -> SomeNode -> Scan`,
This would become
`Project -> Repartition -> SomeNode -> Project -> Repartition -> SomeNode -> Scan`,

Please feel free to close this issue if it seems hypothetical or not-possible

Reply:
Alternately, can we look for just logical relation and replace it with repartition?
```scala
        val shuffled = complementIndexPlan transformUp {
          case relation: LogicalRelation =>
            RepartitionByExpression(attrs or index.indexedColumns, relation, index.numBuckets))
        }
```
(may have missed

Reply:
For Filter Rule, the input plan of `getReplacementPlan` is `Project => Filter => Scan` or `Filter => Scan`.
But For Join Rule, it can be any `linear` plan. 

Since `Repartition => Relation` doesn't support any push-down filters, we need something like `ExtractFilterNode` for better perf. I'll try this.